lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -3853,18 +3853,31 @@ static void materializeAdjointIndirectHelper(AdjointValue val,
                                 destBufferAccess, loc, builder, context);
     }
     break;
-  /// Given a `%buf : *(T0, T1, T2, ...)`, recursively emit instructions to
-  /// materialize the symbolic tuple, filling the buffer.
+  /// Given a `%buf : *(T0, T1, T2, ...)` or `%buf : *Struct` recursively emit
+  /// instructions to materialize the symbolic tuple or struct, filling the
+  /// buffer.
   case AdjointValue::Kind::Aggregate: {
-    SmallVector<SILValue, 8> elements;
-    for (auto eltAndIdx : enumerate(val.getAggregateElements())) {
-      auto idx = eltAndIdx.index();
-      auto *tupTy = val.getSwiftType()->castTo<TupleType>();
-      auto eltTy = SILType::getPrimitiveObjectType(tupTy->getCanonicalType());
-      auto *eltBuf =
-          builder.createTupleElementAddr(loc, destBufferAccess, idx, eltTy);
-      materializeAdjointIndirectHelper(
-          eltAndIdx.value(), eltBuf, builder, context);
+    if (auto *tupTy = val.getSwiftType()->getAs<TupleType>()) {
+      for (auto eltAndIdx : enumerate(val.getAggregateElements())) {
+        auto idx = eltAndIdx.index();
+        auto eltTy = SILType::getPrimitiveObjectType(tupTy->getCanonicalType());
+        auto *eltBuf =
+            builder.createTupleElementAddr(loc, destBufferAccess, idx, eltTy);
+        materializeAdjointIndirectHelper(
+            eltAndIdx.value(), eltBuf, builder, context);
+      }
+    } else if (auto *structDecl = dyn_cast_or_null<StructDecl>(
+                  val.getSwiftType()->getAnyNominal())) {
```

Comment:
You can use `val.getSwiftType().getStructOrBoundGenericStructDecl()` here

Reply:
There are a bunch of other places in this file that use `dyn_cast_or_null<StructDecl>(...getAnyNominal())`, so I think I should keep it this way for consistency. Though I could create a separate PR changing all of them if that is indeed the right thing.

Reply:
> Though I could create a separate PR changing all of them if that is indeed the right thing.

Please do in a separate patch. By iterating over stored properties of a `StructDecl` you'll end up getting unbound generic parameter types. `getStructOrBoundGenericStruct()` handles both bound generic structs and unbound generic structs, and you can directly use the `subst` method to perform substitution.

Reply:
Created https://bugs.swift.org/browse/SR-8770 to make sure I remember.